src/whitenoise/database/accounts_groups.rs

```diff
@@ -0,0 +1,510 @@
+             WHERE account_pubkey = ? AND mls_group_id = ?",
+        )
+        .bind(account_pubkey.to_hex())
+        .bind(mls_group_id.as_slice())
+        .fetch_optional(&database.pool)
+        .await?;
+
+        match row {
+            Some(r) => Ok(Some(r.into_account_group()?)),
+            None => Ok(None),
+        }
+    }
+
+    /// Finds or creates an AccountGroup for the given account and group.
+    /// Returns the AccountGroup and a boolean indicating if it was newly created.
+    ///
+    /// This uses an insert-first approach to avoid TOCTOU race conditions:
+    /// - Attempts to insert first
+    /// - On unique constraint violation, fetches the existing record
+    pub(crate) async fn find_or_create(
+        account_pubkey: &PublicKey,
+        mls_group_id: &GroupId,
+        database: &Database,
+    ) -> Result<(Self, bool), sqlx::Error> {
+        // Try to insert first - this handles the race condition properly
+        match Self::insert_new(account_pubkey, mls_group_id, database).await {
+            Ok(created) => Ok((created, true)),
+            Err(sqlx::Error::Database(db_err)) if db_err.is_unique_violation() => {
+                // Another task inserted the record between our check and insert
+                // Fetch the existing record
+                let existing =
+                    Self::find_by_account_and_group(account_pubkey, mls_group_id, database)
+                        .await?
+                        .ok_or(sqlx::Error::RowNotFound)?;
+                Ok((existing, false))
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Finds all AccountGroups for a given account.
+    pub(crate) async fn find_all_for_account(
+        account_pubkey: &PublicKey,
+        database: &Database,
+    ) -> Result<Vec<Self>, sqlx::Error> {
+        let rows = sqlx::query_as::<_, AccountGroupRow>(
+            "SELECT id, account_pubkey, mls_group_id, user_confirmation, created_at, updated_at
+             FROM accounts_groups
+             WHERE account_pubkey = ?
+             ORDER BY created_at DESC",
+        )
+        .bind(account_pubkey.to_hex())
+        .fetch_all(&database.pool)
+        .await?;
+
+        rows.into_iter()
+            .map(|r| r.into_account_group())
+            .collect::<Result<Vec<_>, _>>()
+    }
+
+    /// Finds all visible AccountGroups for a given account.
+    /// Visible means: user_confirmation is NULL (pending) or true (accepted).
+    /// Declined groups (user_confirmation = false) are hidden.
+    pub(crate) async fn find_visible_for_account(
+        account_pubkey: &PublicKey,
+        database: &Database,
+    ) -> Result<Vec<Self>, sqlx::Error> {
+        let rows = sqlx::query_as::<_, AccountGroupRow>(
+            "SELECT id, account_pubkey, mls_group_id, user_confirmation, created_at, updated_at
+             FROM accounts_groups
+             WHERE account_pubkey = ? AND (user_confirmation IS NULL OR user_confirmation = 1)
```

Comment:
if possible I'd use `TRUE` instead of `1` (please verify that it works though 🙏)